File: apps/android/app/src/main/java/ca/penny/offline/FinanceModels.kt
package ca.penny.offline

import org.json.JSONArray
import org.json.JSONObject
import java.time.LocalDate
import java.time.YearMonth

object FinanceWire {
    fun obj(vararg values: Pair<String, Any?>) = JSONObject().apply { values.forEach { put(it.first, it.second ?: JSONObject.NULL) } }
    fun text(value: String, max: Int = 4000, name: Boolean = false) { Wire.requireUnicode(value); require(value.codePointCount(0,value.length) <= max && (!name || value.isNotEmpty() && value == Wire.trim(value))) { "Check the text length and surrounding spaces" } }
    fun money(value: Long, zero: Boolean = false) { require(value in (if(zero) 0L else 1L)..Money.maxExpense) { "Amount is outside the supported range" } }
    fun month(value: String) { require(Regex("[0-9]{4}-[0-9]{2}").matches(value)); Wire.requireDate("$value-01") }
    fun times(created: String, updated: String) { Wire.requireInstant(created); Wire.requireInstant(updated); require(updated >= created) }
    fun bool(json: JSONObject, key: String) = (json.get(key) as? Boolean) ?: error("Invalid boolean")
    fun nullString(json: JSONObject, key: String): String? = if(json.get(key) === JSONObject.NULL) null else Wire.string(json,key)
    fun nullLong(json: JSONObject, key: String): Long? = if(json.get(key) === JSONObject.NULL) null else Wire.integer(json,key)
    fun sum(values: Iterable<Long>) = values.fold(0L) { total, value -> Math.addExact(total,value).also { require(it <= Money.maxAggregate) { "Total exceeds supported range" } } }
}

interface FinanceRecord { val id: String; fun json(): JSONObject }

data class Schedule(val frequency: String = "monthly", val startDate: String = LocalDate.now().toString(), val endDate: String? = null, val dayOfMonth: Int? = null) {
    init { require(frequency in listOf("once","weekly","biweekly","monthly","yearly")); Wire.requireDate(startDate); endDate?.let { Wire.requireDate(it); require(it >= startDate) }; require(dayOfMonth == null || dayOfMonth in 1..31); require(frequency in listOf("monthly","yearly") || dayOfMonth == null) }
    fun json() = FinanceWire.obj("frequency" to frequency,"startDate" to startDate,"endDate" to endDate,"dayOfMonth" to dayOfMonth)
    companion object { fun decode(j: JSONObject): Schedule { Wire.exactKeys(j,"frequency","startDate","endDate","dayOfMonth"); val day = FinanceWire.nullLong(j,"dayOfMonth"); require(day == null || day in 1..31); return Schedule(Wire.string(j,"frequency"),Wire.string(j,"startDate"),FinanceWire.nullString(j,"endDate"),day?.toInt()) } }
}

data class Budget(override val id: String = Wire.id(), val category: String = Categories.other, val month: String = YearMonth.now().toString(), val limitMinor: Long, val rollover: Boolean = false, val alertThresholdBps: Int = 8000, val notificationsEnabled: Boolean = false, val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); require(category in Categories.all); FinanceWire.month(month); FinanceWire.money(limitMinor,true); require(alertThresholdBps in 0..10000); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"category" to category,"month" to month,"limitMinor" to limitMinor,"rollover" to rollover,"alertThresholdBps" to alertThresholdBps,"notificationsEnabled" to notificationsEnabled,"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object { fun decode(j: JSONObject): Budget { Wire.exactKeys(j,"id","category","month","limitMinor","rollover","alertThresholdBps","notificationsEnabled","createdAt","updatedAt"); val threshold = Wire.integer(j,"alertThresholdBps"); require(threshold in 0..10000); return Budget(Wire.string(j,"id"),Wire.string(j,"category"),Wire.string(j,"month"),Wire.integer(j,"limitMinor"),FinanceWire.bool(j,"rollover"),threshold.toInt(),FinanceWire.bool(j,"notificationsEnabled"),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) } }
}

data class IncomeSource(override val id: String = Wire.id(), val name: String, val category: String = "other", val grossMinor: Long, val netMinor: Long? = null, val currencyCode: String = "CAD", val taxable: Boolean = true, val isRecurring: Boolean = false, val isActive: Boolean = true, val description: String = "", val schedule: Schedule = Schedule("once"), val lastReceivedAt: String? = null, val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); FinanceWire.text(name,200,true); require(category in categories); FinanceWire.money(grossMinor); netMinor?.let { require(it in 0..grossMinor) }; require(currencyCode == "CAD"); FinanceWire.text(description); lastReceivedAt?.let(Wire::requireInstant); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"name" to name,"category" to category,"grossMinor" to grossMinor,"netMinor" to netMinor,"currencyCode" to currencyCode,"taxable" to taxable,"isRecurring" to isRecurring,"isActive" to isActive,"description" to description,"schedule" to schedule.json(),"lastReceivedAt" to lastReceivedAt,"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object {
        val categories = listOf("salary","freelance","bonus","investment","rental","side_hustle","gift","other")
        fun decode(j: JSONObject): IncomeSource { Wire.exactKeys(j,"id","name","category","grossMinor","netMinor","currencyCode","taxable","isRecurring","isActive","description","schedule","lastReceivedAt","createdAt","updatedAt"); return IncomeSource(Wire.string(j,"id"),Wire.string(j,"name"),Wire.string(j,"category"),Wire.integer(j,"grossMinor"),FinanceWire.nullLong(j,"netMinor"),Wire.string(j,"currencyCode"),FinanceWire.bool(j,"taxable"),FinanceWire.bool(j,"isRecurring"),FinanceWire.bool(j,"isActive"),Wire.string(j,"description"),Schedule.decode(j.getJSONObject("schedule")),FinanceWire.nullString(j,"lastReceivedAt"),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) }
    }
}

data class IncomeEntry(override val id: String = Wire.id(), val sourceId: String, val receivedDate: String = LocalDate.now().toString(), val amountMinor: Long, val currencyCode: String = "CAD", val note: String = "", val occurrenceDate: String? = null, val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); Wire.requireId(sourceId); Wire.requireDate(receivedDate); FinanceWire.money(amountMinor); require(currencyCode == "CAD"); FinanceWire.text(note); occurrenceDate?.let(Wire::requireDate); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"sourceId" to sourceId,"receivedDate" to receivedDate,"amountMinor" to amountMinor,"currencyCode" to currencyCode,"note" to note,"occurrenceDate" to occurrenceDate,"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object { fun decode(j: JSONObject): IncomeEntry { Wire.exactKeys(j,"id","sourceId","receivedDate","amountMinor","currencyCode","note","occurrenceDate","createdAt","updatedAt"); return IncomeEntry(Wire.string(j,"id"),Wire.string(j,"sourceId"),Wire.string(j,"receivedDate"),Wire.integer(j,"amountMinor"),Wire.string(j,"currencyCode"),Wire.string(j,"note"),FinanceWire.nullString(j,"occurrenceDate"),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) } }
}

data class SavingsGoal(override val id: String = Wire.id(), val name: String, val category: String = "custom", val targetMinor: Long, val openingMinor: Long = 0, val monthlyContributionMinor: Long = 0, val currencyCode: String = "CAD", val status: String = "active", val isActive: Boolean = true, val priority: String = "medium", val description: String = "", val emoji: String = "", val startDate: String = LocalDate.now().toString(), val targetDate: String? = null, val achievedDate: String? = null, val lastContributionAt: String? = null, val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); FinanceWire.text(name,200,true); require(category in categories); FinanceWire.money(targetMinor); FinanceWire.money(openingMinor,true); FinanceWire.money(monthlyContributionMinor,true); require(currencyCode == "CAD"); require(status in listOf("active","achieved","paused","cancelled")); require(priority in listOf("low","medium","high","critical")); FinanceWire.text(description); FinanceWire.text(emoji,16); Wire.requireDate(startDate); listOfNotNull(targetDate,achievedDate).forEach { Wire.requireDate(it); require(it >= startDate) }; lastContributionAt?.let(Wire::requireInstant); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"name" to name,"category" to category,"targetMinor" to targetMinor,"openingMinor" to openingMinor,"monthlyContributionMinor" to monthlyContributionMinor,"currencyCode" to currencyCode,"status" to status,"isActive" to isActive,"priority" to priority,"description" to description,"emoji" to emoji,"startDate" to startDate,"targetDate" to targetDate,"achievedDate" to achievedDate,"lastContributionAt" to lastContributionAt,"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object {
        val categories = listOf("emergency_fund","travel","education","health","house_down_payment","car","wedding","retirement","investment","custom")
        fun decode(j: JSONObject): SavingsGoal { Wire.exactKeys(j,"id","name","category","targetMinor","openingMinor","monthlyContributionMinor","currencyCode","status","isActive","priority","description","emoji","startDate","targetDate","achievedDate","lastContributionAt","createdAt","updatedAt"); return SavingsGoal(Wire.string(j,"id"),Wire.string(j,"name"),Wire.string(j,"category"),Wire.integer(j,"targetMinor"),Wire.integer(j,"openingMinor"),Wire.integer(j,"monthlyContributionMinor"),Wire.string(j,"currencyCode"),Wire.string(j,"status"),FinanceWire.bool(j,"isActive"),Wire.string(j,"priority"),Wire.string(j,"description"),Wire.string(j,"emoji"),Wire.string(j,"startDate"),FinanceWire.nullString(j,"targetDate"),FinanceWire.nullString(j,"achievedDate"),FinanceWire.nullString(j,"lastContributionAt"),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) }
    }
}

data class SavingsEntry(override val id: String = Wire.id(), val goalId: String, val goalName: String, val date: String = LocalDate.now().toString(), val amountMinor: Long, val currencyCode: String = "CAD", val contributionType: String = "manual", val source: String = "", val note: String = "", val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); Wire.requireId(goalId); FinanceWire.text(goalName,200,true); Wire.requireDate(date); FinanceWire.money(amountMinor); require(currencyCode == "CAD"); require(contributionType in listOf("manual","auto","from_expense_savings")); FinanceWire.text(source); FinanceWire.text(note); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"goalId" to goalId,"goalName" to goalName,"date" to date,"amountMinor" to amountMinor,"currencyCode" to currencyCode,"contributionType" to contributionType,"source" to source,"note" to note,"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object { fun decode(j: JSONObject): SavingsEntry { Wire.exactKeys(j,"id","goalId","goalName","date","amountMinor","currencyCode","contributionType","source","note","createdAt","updatedAt"); return SavingsEntry(Wire.string(j,"id"),Wire.string(j,"goalId"),Wire.string(j,"goalName"),Wire.string(j,"date"),Wire.integer(j,"amountMinor"),Wire.string(j,"currencyCode"),Wire.string(j,"contributionType"),Wire.string(j,"source"),Wire.string(j,"note"),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) } }
}

data class RecurringExpense(override val id: String = Wire.id(), val merchant: String, val amountMinor: Long, val currencyCode: String = "CAD", val category: String = Categories.other, val description: String = "", val note: String = "", val isActive: Boolean = true, val schedule: Schedule = Schedule(), val createdAt: String = Wire.now(), val updatedAt: String = createdAt): FinanceRecord {
    init { Wire.requireId(id); FinanceWire.text(merchant,200,true); FinanceWire.money(amountMinor); require(currencyCode == "CAD" && category in Categories.all); FinanceWire.text(description); FinanceWire.text(note); FinanceWire.times(createdAt,updatedAt) }
    override fun json() = FinanceWire.obj("id" to id,"merchant" to merchant,"amountMinor" to amountMinor,"currencyCode" to currencyCode,"category" to category,"description" to description,"note" to note,"isActive" to isActive,"schedule" to schedule.json(),"createdAt" to createdAt,"updatedAt" to updatedAt)
    companion object { fun decode(j: JSONObject): RecurringExpense { Wire.exactKeys(j,"id","merchant","amountMinor","currencyCode","category","description","note","isActive","schedule","createdAt","updatedAt"); return RecurringExpense(Wire.string(j,"id"),Wire.string(j,"merchant"),Wire.integer(j,"amountMinor"),Wire.string(j,"currencyCode"),Wire.string(j,"category"),Wire.string(j,"description"),Wire.string(j,"note"),FinanceWire.bool(j,"isActive"),Schedule.decode(j.getJSONObject("schedule")),Wire.string(j,"createdAt"),Wire.string(j,"updatedAt")) } }
}

data class FinanceData(val budgets: List<Budget> = emptyList(), val incomeSources: List<IncomeSource> = emptyList(), val incomeEntries: List<IncomeEntry> = emptyList(), val savingsGoals: List<SavingsGoal> = emptyList(), val savingsEntries: List<SavingsEntry> = emptyList(), val recurringExpenses: List<RecurringExpense> = emptyList()) {
    fun domains(): Map<String,List<FinanceRecord>> = linkedMapOf("budgets" to budgets,"incomeSources" to incomeSources,"incomeEntries" to incomeEntries,"savingsGoals" to savingsGoals,"savingsEntries" to savingsEntries,"recurringExpenses" to recurringExpenses)
    fun validate(expenses: List<Expense>) {
        domains().forEach { (name,rows) -> require(rows.size <= limits.getValue(name) && rows.map { it.id }.toSet().size == rows.size) { "Duplicate records or too many $name" } }
        require(budgets.map { "${it.category}/${it.month}" }.toSet().size == budgets.size) { "A budget already exists for this category and month" }
        require(incomeEntries.all { row -> incomeSources.any { it.id == row.sourceId } } && savingsEntries.all { row -> savingsGoals.any { it.id == row.goalId } }) { "Missing financial record reference" }
        val incomeKeys = incomeEntries.mapNotNull { it.occurrenceDate?.let { date -> "${it.sourceId}/$date" } }; require(incomeKeys.toSet().size == incomeKeys.size) { "Income occurrence was already received" }
        val expenseKeys = expenses.mapNotNull { it.recurringTemplateId?.let { id -> require(recurringExpenses.any { template -> template.id == id }); "$id/${it.recurringOccurrenceDate}" } }; require(expenseKeys.toSet().size == expenseKeys.size) { "Expense occurrence was already posted" }
        listOf(budgets.map { it.limitMinor },incomeSources.map { it.grossMinor },incomeEntries.map { it.amountMinor },savingsGoals.map { it.targetMinor },savingsGoals.map { it.openingMinor },savingsGoals.map { it.monthlyContributionMinor },savingsEntries.map { it.amountMinor },recurringExpenses.map { it.amountMinor }).forEach { FinanceWire.sum(it) }
        savingsGoals.forEach { goal -> FinanceWire.sum(listOf(goal.openingMinor) + savingsEntries.filter { it.goalId == goal.id }.map { it.amountMinor }) }
        FinanceWire.sum(savingsGoals.map { it.openingMinor } + savingsEntries.map { it.amountMinor })
    }
    companion object {
        val limits = linkedMapOf("budgets" to 1200,"incomeSources" to 1000,"incomeEntries" to 10000,"savingsGoals" to 1000,"savingsEntries" to 10000,"recurringExpenses" to 1000)
        fun decode(j: JSONObject): FinanceData {
            fun <T> rows(name: String, decode: (JSONObject)->T): List<T> { val list = j.getJSONArray(name); require(list.length() <= limits.getValue(name)); return (0 until list.length()).map { decode(list.getJSONObject(it)) } }
            return FinanceData(rows("budgets",Budget::decode),rows("incomeSources",IncomeSource::decode),rows("incomeEntries",IncomeEntry::decode),rows("savingsGoals",SavingsGoal::decode),rows("savingsEntries",SavingsEntry::decode),rows("recurringExpenses",RecurringExpense::decode))
        }
    }
}

File: apps/ios/PennyOffline/FinanceModels.swift
import Foundation

enum FinanceValidation {
    // ECMAScript trim whitespace used by the portable reference, including BOM.
    static let wireWhitespace = CharacterSet(charactersIn: "\u{0009}\u{000B}\u{000C}\u{0020}\u{00A0}\u{1680}\u{2000}\u{2001}\u{2002}\u{2003}\u{2004}\u{2005}\u{2006}\u{2007}\u{2008}\u{2009}\u{200A}\u{202F}\u{205F}\u{3000}\u{FEFF}\u{000A}\u{000D}\u{2028}\u{2029}")
    static let maximumTotal: Int64 = 999_999_999_999_999
    static func require(_ condition: Bool) throws { if !condition { throw ExpenseError.invalidFinance } }
    static func uuid(_ value: String) throws { try require(UUID(uuidString: value)?.uuidString.lowercased() == value) }
    static func name(_ value: String) throws { try require(!value.isEmpty && value.unicodeScalars.count <= 200 && value == value.trimmingCharacters(in: wireWhitespace)) }
    static func text(_ value: String) throws { try require(value.unicodeScalars.count <= 4_000) }
    static func money(_ value: Int64, zero: Bool = false) throws { try require(value >= (zero ? 0 : 1) && value <= Money.maximumMinor) }
    static func metadata(_ id: String, _ created: String, _ updated: String) throws {
        try uuid(id); try require(CivilDate.validTimestamp(created) && CivilDate.validTimestamp(updated) && updated >= created)
    }
    static func total(_ values: [Int64]) throws -> Int64 {
        var result: Int64 = 0
        for value in values {
            let (sum, overflow) = result.addingReportingOverflow(value)
            try require(!overflow && sum >= 0 && sum <= maximumTotal); result = sum
        }
        return result
    }
}

struct FinanceSchedule: Codable, Equatable, Sendable {
    static let frequencies = ["once", "weekly", "biweekly", "monthly", "yearly"]
    var frequency: String = "once"
    var startDate: String = CivilDate.string(Date())
    var endDate: String? = nil
    var dayOfMonth: Int? = nil
    static let keys: Set<String> = ["frequency", "startDate", "endDate", "dayOfMonth"]
    enum CodingKeys: String, CodingKey { case frequency, startDate, endDate, dayOfMonth }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(frequency, forKey: .frequency)
        try container.encode(startDate, forKey: .startDate)
        try container.encode(endDate, forKey: .endDate)
        try container.encode(dayOfMonth, forKey: .dayOfMonth)
    }
    func validate() throws {
        try FinanceValidation.require(Self.frequencies.contains(frequency) && CivilDate.validDate(startDate))
        try FinanceValidation.require(endDate == nil || CivilDate.validDate(endDate!) && endDate! >= startDate)
        try FinanceValidation.require(dayOfMonth == nil || (1...31).contains(dayOfMonth!))
        try FinanceValidation.require(["monthly", "yearly"].contains(frequency) || dayOfMonth == nil)
    }
}

struct Budget: Codable, Equatable, Sendable, Identifiable {
    var category: String = Categories.other
    var month: String = String(CivilDate.string(Date()).prefix(7))
    var limitMinor: Int64 = 0
    var rollover: Bool = false
    var alertThresholdBps: Int = 8000
    var notificationsEnabled: Bool = false
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["category", "month", "limitMinor", "rollover", "alertThresholdBps", "notificationsEnabled", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case category, month, limitMinor, rollover, alertThresholdBps, notificationsEnabled, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(category, forKey: .category)
        try container.encode(month, forKey: .month)
        try container.encode(limitMinor, forKey: .limitMinor)
        try container.encode(rollover, forKey: .rollover)
        try container.encode(alertThresholdBps, forKey: .alertThresholdBps)
        try container.encode(notificationsEnabled, forKey: .notificationsEnabled)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.require(Categories.all.contains(category) && GregorianDay.validMonth(month) && (0...10_000).contains(alertThresholdBps))
        try FinanceValidation.money(limitMinor, zero: true)
    }
}

struct IncomeSource: Codable, Equatable, Sendable, Identifiable {
    static let categories = ["salary", "freelance", "bonus", "investment", "rental", "side_hustle", "gift", "other"]
    var name: String = ""
    var category: String = "other"
    var grossMinor: Int64 = 1
    var netMinor: Int64? = nil
    var currencyCode: String = "CAD"
    var taxable: Bool = true
    var isRecurring: Bool = false
    var isActive: Bool = true
    var description: String = ""
    var schedule: FinanceSchedule = FinanceSchedule()
    var lastReceivedAt: String? = nil
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["name", "category", "grossMinor", "netMinor", "currencyCode", "taxable", "isRecurring", "isActive", "description", "schedule", "lastReceivedAt", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case name, category, grossMinor, netMinor, currencyCode, taxable, isRecurring, isActive, description, schedule, lastReceivedAt, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(category, forKey: .category)
        try container.encode(grossMinor, forKey: .grossMinor)
        try container.encode(netMinor, forKey: .netMinor)
        try container.encode(currencyCode, forKey: .currencyCode)
        try container.encode(taxable, forKey: .taxable)
        try container.encode(isRecurring, forKey: .isRecurring)
        try container.encode(isActive, forKey: .isActive)
        try container.encode(description, forKey: .description)
        try container.encode(schedule, forKey: .schedule)
        try container.encode(lastReceivedAt, forKey: .lastReceivedAt)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.name(name)
        try FinanceValidation.require(Self.categories.contains(category) && currencyCode == "CAD")
        try FinanceValidation.money(grossMinor)
        if let netMinor { try FinanceValidation.money(netMinor, zero: true); try FinanceValidation.require(netMinor <= grossMinor) }
        try FinanceValidation.text(description)
        try schedule.validate()
        if let lastReceivedAt { try FinanceValidation.require(CivilDate.validTimestamp(lastReceivedAt)) }
    }
}

struct IncomeEntry: Codable, Equatable, Sendable, Identifiable {
    var sourceId: String = ""
    var receivedDate: String = CivilDate.string(Date())
    var amountMinor: Int64 = 1
    var currencyCode: String = "CAD"
    var note: String = ""
    var occurrenceDate: String? = nil
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["sourceId", "receivedDate", "amountMinor", "currencyCode", "note", "occurrenceDate", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case sourceId, receivedDate, amountMinor, currencyCode, note, occurrenceDate, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(sourceId, forKey: .sourceId)
        try container.encode(receivedDate, forKey: .receivedDate)
        try container.encode(amountMinor, forKey: .amountMinor)
        try container.encode(currencyCode, forKey: .currencyCode)
        try container.encode(note, forKey: .note)
        try container.encode(occurrenceDate, forKey: .occurrenceDate)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.uuid(sourceId)
        try FinanceValidation.require(CivilDate.validDate(receivedDate) && currencyCode == "CAD")
        try FinanceValidation.money(amountMinor)
        try FinanceValidation.text(note)
        if let occurrenceDate { try FinanceValidation.require(CivilDate.validDate(occurrenceDate)) }
    }
}

struct SavingsGoal: Codable, Equatable, Sendable, Identifiable {
    static let categories = ["emergency_fund", "travel", "education", "health", "house_down_payment", "car", "wedding", "retirement", "investment", "custom"]
    static let statuses = ["active", "achieved", "paused", "cancelled"]
    static let priorities = ["low", "medium", "high", "critical"]
    var name: String = ""
    var category: String = "custom"
    var targetMinor: Int64 = 1
    var openingMinor: Int64 = 0
    var monthlyContributionMinor: Int64 = 0
    var currencyCode: String = "CAD"
    var status: String = "active"
    var isActive: Bool = true
    var priority: String = "medium"
    var description: String = ""
    var emoji: String = ""
    var startDate: String = CivilDate.string(Date())
    var targetDate: String? = nil
    var achievedDate: String? = nil
    var lastContributionAt: String? = nil
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["name", "category", "targetMinor", "openingMinor", "monthlyContributionMinor", "currencyCode", "status", "isActive", "priority", "description", "emoji", "startDate", "targetDate", "achievedDate", "lastContributionAt", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case name, category, targetMinor, openingMinor, monthlyContributionMinor, currencyCode, status, isActive, priority, description, emoji, startDate, targetDate, achievedDate, lastContributionAt, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(category, forKey: .category)
        try container.encode(targetMinor, forKey: .targetMinor)
        try container.encode(openingMinor, forKey: .openingMinor)
        try container.encode(monthlyContributionMinor, forKey: .monthlyContributionMinor)
        try container.encode(currencyCode, forKey: .currencyCode)
        try container.encode(status, forKey: .status)
        try container.encode(isActive, forKey: .isActive)
        try container.encode(priority, forKey: .priority)
        try container.encode(description, forKey: .description)
        try container.encode(emoji, forKey: .emoji)
        try container.encode(startDate, forKey: .startDate)
        try container.encode(targetDate, forKey: .targetDate)
        try container.encode(achievedDate, forKey: .achievedDate)
        try container.encode(lastContributionAt, forKey: .lastContributionAt)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.name(name)
        try FinanceValidation.require(Self.categories.contains(category) && Self.statuses.contains(status) && Self.priorities.contains(priority) && currencyCode == "CAD")
        try FinanceValidation.money(targetMinor)
        try FinanceValidation.money(openingMinor, zero: true)
        try FinanceValidation.money(monthlyContributionMinor, zero: true)
        try FinanceValidation.text(description)
        try FinanceValidation.require(emoji.unicodeScalars.count <= 16 && CivilDate.validDate(startDate))
        for date in [targetDate, achievedDate].compactMap({ $0 }) { try FinanceValidation.require(CivilDate.validDate(date) && date >= startDate) }
        if let lastContributionAt { try FinanceValidation.require(CivilDate.validTimestamp(lastContributionAt)) }
    }
}

struct SavingsEntry: Codable, Equatable, Sendable, Identifiable {
    static let types = ["manual", "auto", "from_expense_savings"]
    var goalId: String = ""
    var goalName: String = ""
    var date: String = CivilDate.string(Date())
    var amountMinor: Int64 = 1
    var currencyCode: String = "CAD"
    var contributionType: String = "manual"
    var source: String = ""
    var note: String = ""
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["goalId", "goalName", "date", "amountMinor", "currencyCode", "contributionType", "source", "note", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case goalId, goalName, date, amountMinor, currencyCode, contributionType, source, note, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(goalId, forKey: .goalId)
        try container.encode(goalName, forKey: .goalName)
        try container.encode(date, forKey: .date)
        try container.encode(amountMinor, forKey: .amountMinor)
        try container.encode(currencyCode, forKey: .currencyCode)
        try container.encode(contributionType, forKey: .contributionType)
        try container.encode(source, forKey: .source)
        try container.encode(note, forKey: .note)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.uuid(goalId)
        try FinanceValidation.name(goalName)
        try FinanceValidation.require(CivilDate.validDate(date) && currencyCode == "CAD" && Self.types.contains(contributionType))
        try FinanceValidation.money(amountMinor)
        try FinanceValidation.text(source); try FinanceValidation.text(note)
    }
}

struct RecurringExpense: Codable, Equatable, Sendable, Identifiable {
    var merchant: String = ""
    var amountMinor: Int64 = 1
    var currencyCode: String = "CAD"
    var category: String = Categories.other
    var description: String = ""
    var note: String = ""
    var isActive: Bool = true
    var schedule: FinanceSchedule = FinanceSchedule()
    var id: String = UUID().uuidString.lowercased()
    var createdAt: String = CivilDate.timestamp()
    var updatedAt: String = CivilDate.timestamp()
    static let keys: Set<String> = ["merchant", "amountMinor", "currencyCode", "category", "description", "note", "isActive", "schedule", "id", "createdAt", "updatedAt"]
    enum CodingKeys: String, CodingKey { case merchant, amountMinor, currencyCode, category, description, note, isActive, schedule, id, createdAt, updatedAt }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(merchant, forKey: .merchant)
        try container.encode(amountMinor, forKey: .amountMinor)
        try container.encode(currencyCode, forKey: .currencyCode)
        try container.encode(category, forKey: .category)
        try container.encode(description, forKey: .description)
        try container.encode(note, forKey: .note)
        try container.encode(isActive, forKey: .isActive)
        try container.encode(schedule, forKey: .schedule)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
    }
    func validate() throws {
        try FinanceValidation.metadata(id, createdAt, updatedAt)
        try FinanceValidation.name(merchant)
        try FinanceValidation.require(currencyCode == "CAD" && Categories.all.contains(category))
        try FinanceValidation.money(amountMinor)
        try FinanceValidation.text(description); try FinanceValidation.text(note)
        try schedule.validate()
    }
}
